Make DBusGateway bus specific

DunderRoffe opened the pull request:
Rework of DBusGateway and DBusGatewayParser making them agnostic to
configuration for types they are not configured for.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

Commit:
Make DBusGateway bus specific

Rework of DBusGateway and DBusGatewayParser making them agnostic to
configuration for types they are not configured for.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

kursatkobya commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -33,8 +33,19 @@ DBusGateway::DBusGateway(ProxyType type
+    m_totConfig = json_object();
+    m_busConfig = json_array();
+
+    typeStr = m_type == SessionProxy ? SESSION_CONFIG : SYSTEM_CONFIG;
```
kinda hard to read can you at least add parentheses for sides of equiality operator

DunderRoffe replied:
I'll do something about it

breakreturn commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -44,12 +55,12 @@ DBusGateway::~DBusGateway()
     DBusGatewayParser parser;
 
-    if (isError(parser.parseDBusConfigElement(element, m_sessionBusConfig, m_systemBusConfig))) {
+    // TODO: This should really be done with exceptions instead and json_t* as return type.
```
There's an exception class in gateway.h used by the base class nowadays. Perhaps extend that and use here as you'd like. Or leave it to a separate commit.

DunderRoffe replied:
I was thinking of doing that but hesitated in anticipation of the exception design discussion. But sure I can fix this if you like.

breakreturn replied:
Do as you wish, it might be good to wait for that discussion as well if you feel like that.

kursatkobya commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -33,8 +33,19 @@ DBusGateway::DBusGateway(ProxyType type
+    json_object_set(m_totConfig, typeStr, m_busConfig);
+
+    // Send an empty array as config for the irrelevant proxy type
+    const char* unusedTypeStr = m_type == SystemProxy ? SESSION_CONFIG : SYSTEM_CONFIG;
```
also here

kursatkobya commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.h:
```diff
@@ -82,6 +85,12 @@ class DBusGateway :
 private:
+
+    // JSON stuff. These are members to allow continous appends to the config
+    json_t *m_totConfig;
```
can you find better name for this variable i don't think its explicit enough. Suggestion : entireConfig

kursatkobya commented on libsoftwarecontainer/src/gateway/dbus/dbusgatewayparser.cpp:
```diff
@@ -19,45 +19,25 @@
-
-    if (isError(parseDBusBusConfig(element, SESSION_CONFIG, session))) {
+    json_t *configExists = json_object_get(element, key);
+    if (configExists == nullptr) {
```
its not that important but, (nullptr == configExists) would be better her imo.

tobsan replied:
The name `configExists` is not very nice either, use `config` or something similar instead.

kursatkobya reviewed (changes requested):
Generally looks good. I've got just minor comments.

tobsan commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -76,29 +87,26 @@ bool DBusGateway::activateGateway()
-
-    char *config_c = json_dumps(jsonConfig, JSON_COMPACT);
+    // Dump to string and clean up
+    char *config_c = json_dumps(m_totConfig, JSON_COMPACT);
```
You can use implicit conversion to std::string by declaring `std::string config` here directly, then you won't need to `to_string` it.

tobsan commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.h:
```diff
@@ -82,6 +85,12 @@ class DBusGateway :
 
 private:
+
+    // JSON stuff. These are members to allow continous appends to the config
```
continuous*

tobsan reviewed (changes requested):
See inline comments, but I have two additional comments:

1) With this code, the proxy will still be started as long as one provides an array for the respective gateway (system or session).

2) This will still generate errors when one of the gateways are not configured properly, even if one only wants to configure, say, the session bus.

Commit:
Introduce proxy class for DBusGateway instances

Rename old DBusGateway to DBusGatewayInstance, making it specific for the
session or system bus.

Add new class called DBusGateway that acts as a proxy for dbus holding two
instances. One instance for the session bus and one for the sytem bus.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

tobsan commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -18,191 +17,100 @@
+         log_error() << err;
+         throw GatewayError(err);
+     }
+     return ReturnCode::SUCCESS;
```
This is a bit backwards imho. Can't you instead check if the specific bus object was configured and then expect activation to work?

DunderRoffe replied:
Ill look in to it

tobsan commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -18,191 +17,100 @@
+     if (isError(sessionBusParseResult) && isError(systemBusParseResult)) {
+         log_error() << "Neither session nor  system bus could use the given config";
+         return ReturnCode::FAILURE;
+     }
```
I actually would rather see that instead of checking if we had two errors, did a bit of inspection on the config and expect config to work for the given keys. I realize that this would break the design of this container-style class, so it is perhaps not necessary.

DunderRoffe replied:
This would be nice and could be done rather nice without breaking the design. However, it would require the setConfig() to take json_t* as a parameter to not force reimplementation / duplication of the string -> json_t* code.

tobsan commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -18,191 +17,100 @@
+         log_error() << err;
+         throw GatewayError(err);
+     }
+     return ReturnCode::SUCCESS;
```
And the same goes here. Perhaps we can check the gateway state here also before calling teardown so that we know if it is actually needed? That would still work with the container pattern since all gateways (or sub-objects) implements that method.

tobsan commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.h:
```diff
@@ -21,98 +20,50 @@
-
-    // Socket used for exposing D-Bus in container
-    std::string m_socket;
+    virtual ReturnCode readConfigElement(const json_t *) override { return ReturnCode::SUCCESS; }
```
Since you redefine readConfigElement in the cpp file anyway, why have the default implementation like this?

DunderRoffe replied:
I do not define it in the cpp file

tobsan replied:
My bad, I saw the instance class, sorry.

tobsan commented on libsoftwarecontainer/src/gateway/dbus/dbusgatewayinstance.cpp:
```diff
@@ -0,0 +1,225 @@
+{
+    if (isActivated()) {
+        teardown();
+    }
```
I would prefer that members are decref:ed / deleted here, rather than in teardown()

DunderRoffe replied:
That is a bad idea since the config will be saved or memory will be leaked when tearing down without deleting the Gateway.

DunderRoffe commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -18,191 +17,100 @@
+         log_error() << err;
+         throw GatewayError(err);
+     }
+     return ReturnCode::SUCCESS;
```
No since not being configured at activation will result in an exception being thrown.

tobsan commented on libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp:
```diff
@@ -18,191 +17,99 @@
-
-    return true;
+     ReturnCode sessionBusTeardownResult;
+     if (sessionBus.isConfigured()) {
```
I think you meant to use `isActivated()` instead.

kursatkobya commented on libsoftwarecontainer/src/gateway/dbus/dbusgatewayinstance.cpp:
```diff
@@ -0,0 +1,225 @@
+                       , const std::string &gatewayDir
+                       , const std::string &name)
+    : Gateway(ID)
+    , m_type(type)
```
could you please correct initializer list format. I guess it is aggregriated as 
```
(...  , const std::string &name) : 
    Gateway(ID),
    m_type(type)
```

kursatkobya commented on libsoftwarecontainer/src/gateway/dbus/dbusgatewayinstance.cpp:
```diff
@@ -0,0 +1,225 @@
+    ssize_t written = write(m_infp, config.c_str(), count);
+
+    // writing didn't work at all
+    if (written == -1) {
```
`-1 == written` would be better

breakreturn commented on libsoftwarecontainer/src/gateway/dbus/dbusgatewayinstance.cpp:
```diff
@@ -0,0 +1,225 @@
+                                        | Glib::SPAWN_DO_NOT_REAP_CHILD // Lets us do waitpid
+                                    , sigc::slot<void>() // child setup
+                                    , &m_pid
+                                    , &m_infp // stdin
```
We don't put commas like this anymore, they should be at the end of the lines.

kursatkobya reviewed (approved):
small comments, but looks good

breakreturn commented:
This looks good to me given that it has been reviewed while developing it. So apart from the discussion about making sure we catch errors when one of the "instances" can't be configured/activated properly, this is OK to merge for me.

Commit:
Adds exceptions to the dbus parser

Since the dbus gateway is behind a container-pattern gateway, we need to
distinguish between regular config errors and fatal errors. By using
exceptions, we can let them be caught higher up in the hierarchy.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>
